src/core/common/settings.hpp

```diff
@@ -100,6 +100,7 @@ class SettingsBase : public InstanceLocator
         uint32_t        mPreviousPartitionId;    ///< PartitionId
         Mac::ExtAddress mExtAddress;             ///< Extended Address
         uint8_t         mMlIid[OT_IP6_IID_SIZE]; ///< IID from ML-EID
+        uint16_t        mVersion;                ///< Version
```

Comment:
I think adding field to `NetworkInfo` struct  without a behavior change (like from https://github.com/openthread/openthread/pull/4499) would cause the read from previously written settings (by a previous version i.e. version without this change) to fail.

We have same situation with `ParentInfo` and `ChildInfo`, but I guess they are less critical. If we cannot read `NetworkInfo` device would not be able to connect back to network.

We need to follow

>      * 3. To support forward compatibility (rolling forward to a newer
>      *    software version), code reading and processing key values MUST
>      *    process key values that have shorter length.

@bukepo @jwhui  thoughts?

Reply:
According to the latest discussion, changing version should do full reattach. However, I just realized this new field doesn't mean the version is changed(i.e. users can still configure the version to 1.1 after this PR is merged). I will update the code so that only reattach when Thread version is really changed.

Reply:
I will wait the consensus of #4499.

Reply:
@bukepo actually I think it may work out fine. Adding the field would cause in the very first run (with new firmware) not being able to read previously saved `NetworkInfo`, but we would still have Active Dataset. I think we would attach again with a new (randomly generated) extended address (I guess this is fine and we show up as a new node in network). One concern is that we won't recover our previous `mDeviceMode` (which user may have set on device last) but we can require next layers to set it again in such a case?

Reply:
@abtink Yes, I think it can reattach successfully.  Even a user use source code after this PR is merged, it can still configure the Thread version to 1.1. In such case, the version is not changed, is it necessary to reattach? @jwhui

Reply:
This issue should be resolved by #4499.